README.md
## marketeR [![CRAN](http://www.r-pkg.org/badges/version/marketeR)](http://cran.rstudio.com/package=marketeR) [![Downloads](http://cranlogs.r-pkg.org/badges/marketeR?color=brightgreen)](http://www.r-pkg.org/pkg/marketeR)[![License](http://img.shields.io/badge/license-GPL%20%28%3E=%202%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-2.0.html) 

![marketer_logo](https://github.com/fmikaelian/marketeR/blob/master/img/marketer_logo.png?raw=true)

### Enhanced analytics for marketers navigating the ocean of web data :ocean:

The [marketeR package](http://cran.r-project.org/package=marketeR) provides a web analytics toolbox for marketers using services such as Google Analytics, Facebook Insights, etc.

It gives access to essential tricks digital marketers need in their daily work : automatic reports, audience forecasts, performance evaluation, quick data exploration dashboards, and more.

As the swordfish, the marketeR package will provide you the great speed and agility to navigate the ocean of web data :fish: !

### Table of contents

* [Installation](#installation)
  * [From CRAN](#from-cran)
  * [From GitHub](#from-github)
* [Authentification](#authentification)
* [Examples](#examples)
  * [The AutoReport function](#the-autoreport-function)
  * [The PerformanceForecast function](#the-performanceforecast-function)
  * [The WeekSummary function](#the-weeksummary-function)
  * [The FacebookInsightsDashboard function](#the-facebookinsightsdashboard-function)
* [Terms & Conditions](#terms--conditions)
  * [Author](#author)
  * [Copyrights & licenses](#copyrights--licenses)
  * [Dependencies](#dependencies)
  * [Contributing](#contributing)

### Installation

##### From CRAN

Released and tested versions of the marketeR package are available directly via the
[CRAN](http://cran.r-project.org) network, and can be installed from within R via the following code :

```
install.packages("marketeR")
```

##### From GitHub

The last updates of the marketeR package are available via Github, and can be installed from within R via the following code :

```
# install.packages("devtools")
install_github("fmikaelian/marketeR")
```

### Authentification

The marketeR package is using the [Google Analytics Core Reporting API](https://developers.google.com/analytics/devguides/reporting/core/v3/) to get the data you need, and require you to be authentified with an access token. If you don't know what this is about, please follow the instructions below :

- Go to [Google Developers Console](https://console.developers.google.com/project).
- Click on the *Create Project* button to create a new project.
- Click on the *APIs* tab in the sidebar, search for *Analytics API*, and make sure it is correctly enabled.
- Click on the *Credentials* tab in the sidebar,  then on the *add credentials* button, then on the *Oauth 2.0 client ID* option, check the *Other* box, and finally click on the *Create* button.
- You should now be able to get your *Client ID* and *Client secret*.

Now, paste your credentials from within R as shown below :

```
client.id <- "PASTE HERE YOUR CLIENT ID"
client.secret <- "PASTE HERE YOUR CLIENT SECRET"
```

Create your token object, then validate it :

```
token <- Auth(client.id, client.secret)
ValidateToken(token)
```

You can also save your token for future sessions :

```
save(token, file="./token_file")
```

So that in future sessions, you can just load your token and validate it :

```
load("./token_file")
ValidateToken(token)
```

### Examples

Before starting to go deeper into marketeR tricks, remember to make sure that :

- You are authentified (see the **Authentification** section above). 
- You downloaded the complete documentation of the marketeR package, available via the [CRAN](http://cran.r-project.org/package=marketeR) network.

#### The AutoReport function

> **Problem :** You want to get the audience report of your website during the last month.

![PerformanceForecast](https://github.com/fmikaelian/marketeR/blob/master/img/AutoReport.png?raw=true)

The `AutoReport` function is a quick way to check in details how your website performed during the last month. 

The example below shows you how to get the report of your website (*table.id*), during the month of july 2015 (*end.date*). Note that the report will also display the audience evolution since the month of January 2015 (*start.date*).

```
AutoReport(start.date = "2015-01-01", end.date = "2015-07-30", table.id = "ga:XXXXXXX")
```

Once the process is ended, you will find the report as a shareable HTML file in your current working directory. If you are not sure about your current directory location, you can type from within R the following code to figure it out : 

```
getwd()
```

#### The PerformanceForecast function

> **Problem :** You want to know how your website audience will evolve in the next months.

![PerformanceForecast](https://github.com/fmikaelian/marketeR/blob/master/img/PerformanceForecast.png?raw=true)

The `PerformanceForecast` function is a quick way to predict how your website will perform in the next 12 months based on your past scores.

The example below shows you how to get the forecast of your website (*table.id*), in terms of sessions (*metrics*), based on a chosen period of historical data from February 2007 to July 2015 (*start.date/end.date*). Please note that the more historical data you have, the more accurate the forecast will be.

```
PerformanceForecast(start.date = "2007-02-01", end.date ="2015-06-30", metrics = "ga:sessions", table.id = "ga:XXXXXXX", export = TRUE)
```

Once the process is ended, as the *export* option is set as "TRUE", both raw data & graphics will be exported in the current working directory. If the *export* option is set as "FALSE", R will only print raw data and display a visualization of the forecast.

#### The WeekSummary function

> **Problem :** You want to know how your website audience performed during the last 7 days, compared to your usual audience scores.

![WeekSummary](https://github.com/fmikaelian/marketeR/blob/master/img/WeekSummary.png?raw=true)

The `WeekSummary` function is a quick way to check how your website performed during the past 7 days, compared to your past scores.

The example below shows you how to get the week summary of your website (*table.id*), in terms of sessions (*metrics*).

```
WeekSummary(metrics = "ga:sessions", table.id = "ga:XXXXXXX", export = TRUE)
```

Once the process is ended, as the *export* option is set as "TRUE", both raw data & graphics will be exported in the current working directory. If the *export* option is set as "FALSE", R will only print raw data and display a visualization of the week summary.

#### The FacebookInsightsDashboard function

> **Problem :** You are using Facebook insights in your daily work but you are not happy with its limited functions.

![FacebookInsightsDashboard](https://github.com/fmikaelian/marketeR/blob/master/img/FacebookInsightsDashboard.png?raw=true)

The `FacebookInsightsDashboard` function is a faster and easier way to explore your Facebook data without having to care about coding. 

##### Setting up the dashboard

The FacebookInsightsDashboard set up process is divided into the folowing 4 steps :

###### 1. Access the dashboard

The example below shows you how to acess the dashboard from within R.

```
FacebookInsightsDashboard()
```

###### 2. Get your token

You can get your token by launching the Graph API explorer from Facebook, available [here.](https://developers.facebook.com/tools/explorer/) You just have to copy/paste it from the box shown above to the dashboard.
For security reasons, your token will expire if you log-out your Facebook account or after 30 minutes of activity.

![Token](https://github.com/fmikaelian/marketeR/blob/master/img/Token.png?raw=true)

###### 3. Get your object_ID
Your object_ID could be a page_ID, a post_ID, or a domain_ID. Assuming you want to get insights for a page, you will find the page_ID within the Facebook URL of the page. The URL syntax should look like this : facebook.com/page_ID

###### 4. Select your query parameters
+ *Dates :*
Just select the date range for your insights query.

+ *Metric :*
The insights metric reference is available [here.](https://developers.facebook.com/docs/graph-api/reference/v2.4/insights) As you can see, some metrics are page-related, some others are post-related or even domain-related. So be sure you selected an object_ID type that matches the metric you chose.

+ *Period :*
The period parameter is different from the *start-date / end-date* parameter. For instance, if you chose a period=*week*, results will show daily values. For instance, if the *page_impressions* (period=*week*) for the date Friday May 8th is X, that means that from Friday May 1st to Friday May 8th, there was a total of X *page_impressions*. It is the same concept for other period parameters.

+ *Download buttons :*
If you need to export the graphics generated by the dashboard, or just want to store the raw data extracted from Facebook, you can use the download buttons.


### Terms & Conditions

##### Author

Félix MIKAELIAN <felix.mikaelian@essec.edu> :fr:

##### Copyrights & licenses

- The marketeR package is under license GPL (>= 2).
- The marketeR logo is using the *"Enriqueta"* font by [FontFuror](http://www.fontfuror.com/).
- The marketeR logo is using the icon *"Swordfish"* by [Catalina Montes](http://www.pezglobo.cl/).

##### Dependencies

The marketeR package use the following packages :

- RGoogleAnalytics
- Rfacebook
- forecast
- zoo
- plyr
- dplyr
- ggplot2
- xlsx
- grid
- scales
- shiny
- rmarkdown
- knitr
- ggthemes

##### Contributing

The marketeR package is the compilation of some work tools I've been creating for specific work purposes. I thought that other digital marketers would also find my tricks interesting for their daily work, so I decided to release it as an open-source project.

I would also love ideas to make marketeR better, so feel free to contribute to this exciting project :v: !


NEWS.md
## NEWS

### marketeR 0.1.1 (2015-10-21)

* Bug fixes *

- AutoReport() function : The output html file is now appearing in the working directory.
- WeekSummary() function : The output plot is now showing in the R session.

### marketeR 0.1.0 (2015-09-08)

- Initial release on GitHub
- Initial release on CRAN

R/Settings.R
#-------------------------------------------------------------------------------
# Settings.R (marketeR) - made with <3 by fmikaelian
#-------------------------------------------------------------------------------

#-------------------------------------------------------------------------------
#' @import RGoogleAnalytics Rfacebook forecast zoo dplyr ggplot2 xlsx grid
#'   scales shiny rmarkdown knitr ggthemes
#' @importFrom plyr ddply
#' @importFrom grDevices dev.off png
#' @importFrom graphics plot
#' @importFrom stats setNames ts
#-------------------------------------------------------------------------------

globalVariables(c("token",
                  "as.yearmon",
                  "sessions",
                  "socialNetwork",
                  "yearMonth",
                  "deviceCategory",
                  "percentages",
                  "percentNewsessions",
                  "avgSessionDuration",
                  "bounceRate",
                  "channelGrouping",
                  "userGender",
                  "userAgeBracket",
                  "dayOfWeek",
                  "V2",
                  "dates",
                  "new_dates"
                  )
                )

R/FacebookInsightsDashboard.R
#-------------------------------------------------------------------------------
# FacebookInsightsDashboard.R (marketeR) - made with <3 by fmikaelian
#-------------------------------------------------------------------------------

#-------------------------------------------------------------------------------
#' @title A simple dashboard for social media managers who use
#'   Facebook insights
#' @description \code{FacebookInsightsDashboard} allows faster and easier data
#'   exploration without having to care about coding.
#' @details The dashboard parameters are :
#' \itemize{
#' 
#' \item{\code{token : }} {You can get your token by launching the Graph API explorer from
#'   Facebook, available at \url{https://developers.facebook.com/tools/explorer/}.
#'   You just have to copy/paste it from the box shown above to the dashboard.
#'   For security reasons, your token will expire if you log-out your Facebook
#'   account or after 30 minutes of activity.}
#' \item{\code{object_ID : }} {Your \code{object_ID} could be a \code{page_ID}, a
#'   \code{post_ID}, or a \code{domain_ID}. Assuming you want to get insights for
#'   a page, you will find the \code{page_ID} within the Facebook URL of the page.
#'   The URL syntax should look like this : facebook.com/\code{page_ID}.}
#' \item{\code{dates : }} {Just select the date range for your insights query.}
#' \item{\code{metric :}} {The insights metric reference is available at
#'   \url{https://developers.facebook.com/docs/graph-api/reference/v2.4/insights}.
#'   As you can see, some metrics are page-related, some others are post-related
#'   or even domain-related. So be sure you selected an \code{object_ID} type
#'   that matches the metric you chose.}
#' \item{\code{period : }} {The period parameter is different from the \code{start/end-date}
#'   parameter. For instance, if you chose a \code{period=week}, results will
#'   still show daily values. To be more precise, if the \code{page_impressions}
#'   (\code{period=week}) for the date Friday May 8th is X, that means that from
#'   Friday May 1st to Friday May 8th, there was a total of X
#'   \code{page_impressions}. It is the same concept for other period parameters.}
#' }
#' 
#' @note If you need to export the graphics generated by the dashboard, or just
#'   want to store the raw data extracted from Facebook, you can use the
#'   download buttons.
#' @examples \dontrun{FacebookInsightsDashboard()}
#' @export
#-------------------------------------------------------------------------------

FacebookInsightsDashboard <- function() {

  #-----------------------------------------------------------------------------
  # Find then run the Shiny app
  #-----------------------------------------------------------------------------

  appDir <- system.file("FacebookInsightsDashboard", package = "marketeR")
  if (appDir == "") {
    stop("Could not find given directory. Try re-installing marketeR.", call. = FALSE)
  }

  shiny::runApp(appDir, display.mode = "normal")

}


R/AutoReport.R
#-------------------------------------------------------------------------------
# AutoReport.R (marketeR) - made with <3 by fmikaelian
#-------------------------------------------------------------------------------

#-------------------------------------------------------------------------------
#' @title A detailed report of your website performance during a given month
#' @description \code{Autoreport} is a quick way to check in details how your
#'   website performed during a given month.
#' @param start.date Start date for fetching Analytics data. Requests can
#'   specify a start date formatted as YYYY-MM-DD, or as a relative date
#'   (e.g., today, yesterday, or NdaysAgo where N is a positive integer).
#' @param end.date End date for fetching Analytics data. Request can specify
#'   an end date formatted as YYYY-MM-DD, or as a relative date (e.g., today,
#'   yesterday, or NdaysAgo where N is a positive integer).
#' @param table.id The unique table ID of the form ga:XXXX, where XXXX is the
#'   Analytics view (profile) ID for which the query will retrieve the data.
#' @note The \code{AutoReport} function will generate a shareable HTML file
#'  in your working directory.
#' @examples \dontrun{
#'   AutoReport(start.date = "2015-01-01", end.date = "2015-07-30",
#'   table.id = "ga:XXXXXXX")}
#' @export
#-------------------------------------------------------------------------------

AutoReport <- function(start.date, end.date, table.id) {

  #-----------------------------------------------------------------------------
  # Get report data from Google Analytics Core Reporting API
  #-----------------------------------------------------------------------------

  # basicMetrics.data ----------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:sessions, ga:percentNewsessions,
                                ga:avgSessionDuration, ga:bounceRate",
                     dimensions = "ga:yearMonth",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  basicMetrics.data <- GetReportData(ga.query, token)

  # channelGrouping.data -------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:channelGrouping",
                     sort = "-ga:sessions",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  channelGrouping.data <- GetReportData(ga.query, token)

  # socialNetwork.data ---------------------------------------------------------

  query.list <- Init(start.date = paste(substr(end.date, 1, 8), "01", sep = ""),
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:socialNetwork",
                     sort = "-ga:sessions",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  socialNetwork.data <- GetReportData(ga.query, token)

  # keyword.data ---------------------------------------------------------------

  query.list <- Init(start.date = paste(substr(end.date, 1, 8), "01", sep = ""),
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:keyword",
                     sort = "-ga:sessions",
                     max.results = 10,
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  keyword.data <- GetReportData(ga.query, token)

  # deviceCategory.data --------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:deviceCategory",
                     sort = "-ga:sessions",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  deviceCategory.data <- GetReportData(ga.query, token)

  # topContent.data ------------------------------------------------------------

  query.list <- Init(start.date = paste(substr(end.date, 1, 8), "01", sep = ""),
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:pagePath",
                     sort = "-ga:sessions",
                     max.results = 10,
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  topContent.data <- GetReportData(ga.query, token)

  # country.data ---------------------------------------------------------------

  query.list <- Init(start.date = paste(substr(end.date, 1, 8), "01", sep = ""),
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:country",
                     sort = "-ga:sessions",
                     max.results = 10,
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  country.data <- GetReportData(ga.query, token)

  # userGender.data ------------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:userGender",
                     sort = "-ga:sessions",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  userGender.data <- GetReportData(ga.query, token)

  # userAgeBracket.data --------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:sessions",
                     dimensions = "ga:yearMonth, ga:userAgeBracket",
                     sort = "-ga:sessions",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  userAgeBracket.data <- GetReportData(ga.query, token)

  # newUsers.data --------------------------------------------------------------

  query.list <- Init(start.date = start.date,
                     end.date = end.date,
                     metrics = "ga:newUsers",
                     dimensions = "ga:yearMonth",
                     table.id = table.id)

  ga.query <- QueryBuilder(query.list)
  newUsers.data <- GetReportData(ga.query, token)

  #-----------------------------------------------------------------------------
  # Clean report data
  #-----------------------------------------------------------------------------

  # Change the report data date format -----------------------------------------

  yearmonFormat <- function (x) {
    x <- as.Date(as.yearmon(x, "%Y%m"))
  }

  basicMetrics.data$yearMonth <- yearmonFormat(basicMetrics.data$yearMonth)
  channelGrouping.data$yearMonth <- yearmonFormat(channelGrouping.data$yearMonth)
  socialNetwork.data$yearMonth <- yearmonFormat(socialNetwork.data$yearMonth)
  keyword.data$yearMonth <- yearmonFormat(keyword.data$yearMonth)
  deviceCategory.data$yearMonth <- yearmonFormat(deviceCategory.data$yearMonth)
  topContent.data$yearMonth <- yearmonFormat(topContent.data$yearMonth)
  country.data$yearMonth <- yearmonFormat(country.data$yearMonth)
  userGender.data$yearMonth <- yearmonFormat(userGender.data$yearMonth)
  userAgeBracket.data$yearMonth <- yearmonFormat(userAgeBracket.data$yearMonth)

  # Add percentages ------------------------------------------------------------

  channelGrouping.data <- ddply(channelGrouping.data,
                                "yearMonth",
                                transform,
                                percentages = sessions / sum(sessions))

  deviceCategory.data <- ddply(deviceCategory.data,
                               "yearMonth",
                               transform,
                               percentages = sessions / sum(sessions))

  userGender.data <- ddply(userGender.data,
                           "yearMonth",
                           transform,
                           percentages = sessions / sum(sessions))

  userAgeBracket.data <- ddply(userAgeBracket.data,
                               "yearMonth",
                               transform,
                               percentages = sessions / sum(sessions))

  # Prepare socialNetwork.data data for making a ggplot2 pie chart -------------

  others = data.frame(yearMonth = socialNetwork.data$yearMonth[1],
                      socialNetwork = "Others",
                      sessions = sum(socialNetwork.data$sessions))

  socialNetwork.data <-
    socialNetwork.data %>%
    slice(1:5)

  others$sessions = others$sessions-sum(socialNetwork.data$sessions)
  socialNetwork.data <- bind_rows(socialNetwork.data, others)

  socialNetwork.data <-
    socialNetwork.data %>%
    dplyr::select(socialNetwork, sessions)

  #-----------------------------------------------------------------------------
  # Preparing data for the Executive Summary
  #-----------------------------------------------------------------------------

  ## sessions ------------------------------------------------------------------

  sessions.currentMonth <- as.numeric(basicMetrics.data$sessions[nrow(basicMetrics.data)])
  sessions.lastMonth <- as.numeric(basicMetrics.data$sessions[nrow(basicMetrics.data)-1])
  sessions.relativeChange <- ((sessions.currentMonth/sessions.lastMonth)-1)*100

  ## percentNewsessions --------------------------------------------------------

  percentNewsessions.currentMonth <- as.numeric(basicMetrics.data$percentNewsessions[nrow(basicMetrics.data)])
  percentNewsessions.lastMonth <- as.numeric(basicMetrics.data$percentNewsessions[nrow(basicMetrics.data)-1])
  percentNewsessions.relativeChange <- ((percentNewsessions.currentMonth/percentNewsessions.lastMonth)-1)*100

  ## avgSessionDuration --------------------------------------------------------

  avgSessionDuration.currentMonth <- as.numeric(basicMetrics.data$avgSessionDuration[nrow(basicMetrics.data)])
  avgSessionDuration.lastMonth <- as.numeric(basicMetrics.data$avgSessionDuration[nrow(basicMetrics.data)-1])
  avgSessionDuration.relativeChange <- ((avgSessionDuration.currentMonth/avgSessionDuration.lastMonth)-1)*100

  ## bounceRate ----------------------------------------------------------------

  bounceRate.currentMonth <- as.numeric(basicMetrics.data$bounceRate[nrow(basicMetrics.data)])
  bounceRate.lastMonth <- as.numeric(basicMetrics.data$bounceRate[nrow(basicMetrics.data)-1])
  bounceRate.relativeChange <- ((bounceRate.currentMonth/bounceRate.lastMonth)-1)*100

  ## mobileRate
  deviceCategory.currentMonth <-
    deviceCategory.data %>%
    filter(yearMonth == basicMetrics.data$yearMonth[nrow(basicMetrics.data)]) %>%
    select(deviceCategory, percentages) %>%
    filter(deviceCategory == "mobile")

  deviceCategory.currentMonth$percentages <- as.numeric(substr(deviceCategory.currentMonth$percentages, 0 , nchar(deviceCategory.currentMonth$percentages) - 1))

  deviceCategory.lastMonth <-
    deviceCategory.data %>%
    filter(yearMonth == basicMetrics.data$yearMonth[nrow(basicMetrics.data)-1]) %>%
    select(deviceCategory, percentages) %>%
    filter(deviceCategory == "mobile")

  deviceCategory.lastMonth$percentages <- as.numeric(substr(deviceCategory.lastMonth$percentages, 0 , nchar(deviceCategory.lastMonth$percentages) - 1))

  deviceCategory.relativeChange <- ((deviceCategory.currentMonth$percentages/deviceCategory.lastMonth$percentages)-1)*100

  #-----------------------------------------------------------------------------
  # Create plots of report data
  #-----------------------------------------------------------------------------

  # sessions.plot --------------------------------------------------------------

  sessions.plot <- ggplot(basicMetrics.data, aes(x = yearMonth, y = sessions)) +
    ggtitle("Sessions") +
    geom_line(size = 0.5, color = "red") +
    geom_point(size = 2.5, shape = 21, fill = "white") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = comma, limits=c(0, max(basicMetrics.data$sessions))) +
    theme_economist() +
    scale_colour_economist() +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # percentNewsessions.plot ----------------------------------------------------

  percentNewsessions.plot <-
    ggplot(basicMetrics.data, aes(x = yearMonth, y = percentNewsessions/100)) +
    ggtitle("Percent of New Sessions") +
    geom_line(size = 0.5, color = "red") +
    geom_point(size = 2.5, shape = 21, fill = "white") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels  = percent_format(), limits=c(0, max(basicMetrics.data$percentNewsessions/100))) +
    theme_economist() +
    scale_colour_economist() +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # avgSessionDuration.plot ----------------------------------------------------

  avgSessionDuration.plot <-
    ggplot(basicMetrics.data, aes(x = yearMonth, y = avgSessionDuration)) +
    ggtitle("Average Session Duration (in seconds)") +
    geom_line(size = 0.5, color = "red") +
    geom_point(size = 2.5, shape = 21, fill = "white") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = comma, limits=c(0, max(basicMetrics.data$avgSessionDuration))) +
    theme_economist() +
    scale_colour_economist() +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # bounceRate.plot ----------------------------------------------------

  bounceRate.plot <-
    ggplot(basicMetrics.data, aes(x = yearMonth, y = bounceRate/100)) +
    ggtitle(" Bounce Rate (in percent)") +
    geom_line(size = 0.5, color = "red") +
    geom_point(size = 2.5, shape = 21, fill = "white") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels  = percent_format(), limits=c(0, max(basicMetrics.data$bounceRate/100))) +
    theme_economist() +
    scale_colour_economist() +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # channelGrouping.plot -------------------------------------------------------

  channelGrouping.plot <-
    ggplot(channelGrouping.data, aes(x = yearMonth, y = percentages)) +
    geom_bar(aes(fill = channelGrouping), stat = "identity") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = percent_format()) +
    theme_economist() +
    scale_colour_economist() +
    scale_fill_brewer(palette = "YlGnBu") +
    guides(fill=guide_legend(title = NULL)) +
    theme(legend.text=element_text(size=10),
          axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # socialNetwork.plot ---------------------------------------------------------

  socialNetwork.plot <-
    ggplot(socialNetwork.data, aes(x = "", y = sessions, fill = socialNetwork)) +
    geom_bar(width = 1, stat = "identity") +
    coord_polar("y") +
    theme_economist() +
    scale_fill_brewer(palette = "Pastel2") +
    guides(fill=guide_legend(title = NULL)) +
    geom_text(
      aes(y = sessions/3 + c(0, cumsum(sessions)[-length(sessions)])),
      label = paste(round(socialNetwork.data$sessions / sum(socialNetwork.data$sessions) * 100, 0), "%", sep=""),
      size=8,
      color = "white") +
    theme(legend.position = "right",
          axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          panel.border = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())

  # deviceCategory.plot --------------------------------------------------------

  deviceCategory.plot <-
    ggplot(deviceCategory.data, aes(x = yearMonth, y = percentages)) +
    geom_bar(aes(fill = deviceCategory), stat = "identity") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = percent_format()) +
    theme_economist() +
    scale_colour_economist() +
    scale_fill_brewer(palette = "RdPu") +
    guides(fill=guide_legend(title = NULL)) +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # userGender.plot ------------------------------------------------------------

  userGender.plot <-
    ggplot(userGender.data, aes(x = yearMonth, y = percentages)) +
    geom_bar(aes(fill = userGender), stat = "identity") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = percent_format()) +
    theme_economist() +
    scale_colour_economist() +
    scale_fill_brewer(palette = "Pastel1") +
    guides(fill=guide_legend(title = NULL)) +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  # userAgeBracket.plot --------------------------------------------------------

  userAgeBracket.plot <-
    ggplot(userAgeBracket.data, aes(x = yearMonth, y = percentages)) +
    geom_bar(aes(fill = userAgeBracket), stat = "identity") +
    scale_x_date(labels = date_format("%b %Y"), breaks = date_breaks("month")) +
    scale_y_continuous(labels = percent_format()) +
    theme_economist() +
    scale_colour_economist() +
    scale_fill_brewer(palette = "Pastel1") +
    guides(fill=guide_legend(title = NULL)) +
    theme(axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 10))

  #-----------------------------------------------------------------------------
  # Multiple plot function from cookbook-r.com by Winston Chang
  #-----------------------------------------------------------------------------

  multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {

    # Make a list from the ... arguments and plotlist
    plots <- c(list(...), plotlist)

    numPlots = length(plots)

    # If layout is NULL, then use 'cols' to determine layout
    if (is.null(layout)) {
      # Make the panel
      # ncol: Number of columns of plots
      # nrow: Number of rows needed, calculated from # of cols
      layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                       ncol = cols, nrow = ceiling(numPlots/cols))
    }

    if (numPlots==1) {
      print(plots[[1]])

    } else {
      # Set up the page
      grid.newpage()
      pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

      # Make each plot, in the correct location
      for (i in 1:numPlots) {
        # Get the i,j matrix positions of the regions that contain this subplot
        matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

        print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                        layout.pos.col = matchidx$col))
      }
    }
  }

  #-----------------------------------------------------------------------------
  # Find then render the AutoReport.rmd file
  #-----------------------------------------------------------------------------

  rmdDir <- system.file("AutoReport", "AutoReport.Rmd", package = "marketeR")
  if (rmdDir == "") {
    stop("Could not find given directory. Try re-installing marketeR.", call. = FALSE)
  }

  rmarkdown::render(rmdDir, output_file = '~/AutoReport.html')

}
